integration_tests/src/main/python/data_gen.py

```diff
@@ -613,31 +613,112 @@ def make_null():
         self._start(rand, make_null)
 
 # DayTimeIntervalGen is for Spark 3.3.0+
-# DayTimeIntervalType(startField, endField): Represents a day-time interval which is made up of a contiguous subset of the following fields:
+# DayTimeIntervalType(startField, endField):
+# Represents a day-time interval which is made up of a contiguous subset of the following fields:
 #   SECOND, seconds within minutes and possibly fractions of a second [0..59.999999],
+#   Note Spark now uses 99 as max second, see issue https://issues.apache.org/jira/browse/SPARK-38324
```

Comment:
The finer granularity you mentioned will be trancated,  code is: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala#L1037toL1056

see below:
```
from pyspark.sql.types import *
from datetime import date, datetime, timedelta, timezone
schema = StructType([
    StructField("c1", DayTimeIntervalType(
                DayTimeIntervalType.DAY, DayTimeIntervalType.DAY
            ), False),
])
data = [(timedelta( days=50, seconds=27), ),]
df = spark.createDataFrame(
        SparkContext.getOrCreate().parallelize(data, numSlices=2),
        schema)
df.show()
+-----------------+
|               c1|
+-----------------+
|INTERVAL '50' DAY|
+-----------------+

```

Reply:
- Yes, casting to string truncates by looking at start and end fields, calling the following code
```
  def toDayTimeIntervalString(
      micros: Long,
      style: IntervalStyle,
      startField: Byte,
      endField: Byte): String = {
```

- After debug, I found the internal long value is not truncated when creating `df`, it stores the micros long value.
The `endField` is always `SECOND` in `durationToMicros`, see below:
```
  def durationToMicros(duration: Duration): Long = {
    durationToMicros(duration, DT.SECOND)
  }

  def durationToMicros(duration: Duration, endField: Byte)
```
- In summary:
The external type of `DayTimeInterval` is Duration, it's converted to micros long value regardless of what start and end fields are.
When casting to string, Spark looks at start and end fields to tuncate the micros.
The pyspark `DayTimeIntervalGen` can just pass a random micros.

- The below code will call the following:
`durationToMicros(duration, DT.SECOND)` create `df`,  the interal long is: **micros of 1 day + micros of 1 second**
Cast to string, calls `toDayTimeIntervalString`,  **truncating the 1 second**
```
  test("my test") {
    val data = Seq(Row(Duration.ofDays(1).plusSeconds(1)))
    val schema = StructType(Array(
      StructField("t", DayTimeIntervalType(DayTimeIntervalType.DAY, DayTimeIntervalType.DAY))
    ))
    val df = spark.createDataFrame(spark.sparkContext.parallelize(data), schema)
    df.show()
  }
```

We can simplify the `DayTimeIntervalGen` by only generate a random micros regardless of (start, end) field, see the current implimetation of `DayTimeIntervalGen`

BTW Because of Duration is external type,  the plugin do not need to handle this type, so this issue  https://github.com/NVIDIA/spark-rapids/issues/1948 can be closed.

Reply:
Filed a Spark issue: Interval types are not truncated to the expected endField when creating a DataFrame via Duration https://issues.apache.org/jira/browse/SPARK-38577[](https://github.com/res-life)